### DISCUSSION TITLE
Enable 2FA on cluster with existing users

### DISCUSSION THREAD

**Discussion by kabachook** [gravitational/teleport#13521] [Q&A] [answered]
Hi! I would like to enable `second_factor: otp` in my cluster, but the following problem occurs: users, who doesn't have an active session will be locked since they didn't set up a TOTP. 

Alternatively I tried to provide a grace period for adding TOTP using `second_factor: optional`, however I can't completely disable Webauthn or U2F options (At least one should be enabled). 
I don't want to leave Webauthn or U2F since:
- Webauthn doesn't work on Windows using tsh
- Webauthn set up in web panel doesn't work in tsh on macOS, so that user is unable to use tsh since they can't login

**Comment by webvictim** [comment, marked as answer]
Interesting dilemma. I think you'd need to approach all your users who don't have a second factor configured and ask them to add one via the web UI or `tsh mfa add --type=TOTP` before enabling that setting.

You can probably put something together to iterate over the `local_auth` section in output of `tctl get user --with-secrets --format json` to figure out who has MFA devices associated already.

**Reply by kabachook** [reply]
Thanks! How do you feel about an enroll page for this case? Basically it will be just the standard sign up page, but without username and password fields.
So that flow will be the following:
1. Cluster migrates to `second_factor: on`
2. User logs in with username and password
3. The enroll page is presented and the user is able add TOTP

**Reply by webvictim** [reply]
Nice idea. Please feel free to raise a feature request: https://github.com/gravitational/teleport/issues/new?assignees=&labels=feature-request&template=feature_request.md
